File: dispatch/python/module/scheduling/runnable_dispatcher.py
###########################################################################
# Copyright (C) 2023 ETH Zurich
# CLAID: Closing the Loop on AI & Data Collection (https://claid.ethz.ch)
# Core AI & Digital Biomarker, Acoustic and Inflammatory Biomarkers (ADAMMA)
# Centre for Digital Health Interventions (c4dhi.org)
# 
# Authors: Patrick Langer
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#         http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

import threading
import time
from typing import List
from logger.logger import Logger
from datetime import datetime

class RunnableDispatcher:
    def __init__(self, main_thread_runnables_queue):
        self.scheduled_runnables = []
        self.mutex = threading.Lock()
        self.condition_variable = threading.Condition(self.mutex)
        self.reschedule_required = False
        self.stopped = True
        self.thread = None
        self.main_thread_runnables_queue = main_thread_runnables_queue

    def get_wait_duration_until_next_runnable_is_due(self):
        with self.mutex:
            if not self.scheduled_runnables:
                # Wait forever.
                return float(1000000)

            microseconds_until_next_runnable_is_due = int((
                self.scheduled_runnables[0].schedule.get_execution_time() - datetime.now()
            ).total_seconds()*1000)

            # Make sure we do not return a negative duration.
            return max(0, microseconds_until_next_runnable_is_due)

    def wait_until_runnable_is_due_or_reschedule_is_required(self):
        # If there is no runnable added, this will be infinity.
        # Hence, we will wait forever and wake up if a reschedule is required.
        wait_time_in_seconds = float(self.get_wait_duration_until_next_runnable_is_due() / 1000)

        # The return value of wait_for can be used to determine whether the wait exited because time passed (False),
        # or because the predicate (self.reschedule_required or self.stopped) evaluates to True (True).
        # However, we are not interested in distinguishing the two cases.
        # In any case, when we wake up, we see if we need to execute anything.
        # wait_for will atomically release the mutex and sleep, and will atomically lock the mutex after waiting.
        with self.condition_variable:
            self.condition_variable.wait_for(
                lambda: self.reschedule_required or self.stopped, wait_time_in_seconds
            )

    def process_runnable(self, scheduled_runnable):
        if self.stopped:
            return

        if scheduled_runnable.runnable.is_valid():
            

            # SWitch to the main thread
            self.main_thread_runnables_queue.put(scheduled_runnable)

            if scheduled_runnable.runnable.stop_dispatcher_after_this_runnable:
                # No more runnables will be executed after this one!
                self.stop()
                Logger.log_info("STOPPED DISPATCHER!")
                return

            if scheduled_runnable.schedule.does_runnable_have_to_be_repeated():
                with self.mutex:
                    planned_execution_time = scheduled_runnable.schedule.get_execution_time()
                    scheduled_runnable.schedule.update_execution_time()

                    # planned_execution_time_str = planned_execution_time.strftime(
                    #     "%d.%m.%y %H:%M:%S"
                    # )
                    # current_execution_str = datetime.now().strftime(
                    #     "%d.%m.%y %H:%M:%S"
                    # )

                    # new_execution_time = scheduled_runnable.schedule.get_execution_time()
                    # next_execution_str = new_execution_time.strftime(
                    #     "%d.%m.%y %H:%M:%S"
                    # )

                    # Logger.log_info(
                    #     f"Runnable, scheduled for execution at {planned_execution_time_str}, has been executed at {current_execution_str}, scheduling next execution for {next_execution_str}",
                    # )

                    # Reinsert the runnable with new scheduled execution time.
                    # Note, that the runnable was removed from scheduled_runnables in the get_and_remove_due_runnables() function.
                    self.scheduled_runnables.append(scheduled_runnable)
                    self.scheduled_runnables.sort(
                        key=lambda x: x.schedule.get_execution_time()
                    )

    def process_runnables(self, runnables):
        for idx, runnable in enumerate(runnables):
            self.process_runnable(runnable)

    def get_and_remove_due_runnables(self):
        runnables = []
        now = datetime.now()

        with self.mutex:
            while self.scheduled_runnables:
                due_runnable = self.scheduled_runnables[0]
                if now >= due_runnable.schedule.get_execution_time():
                    runnables.append(due_runnable)
                    self.scheduled_runnables.pop(0)
                else:
                    break

       

        return runnables

    def run_scheduling(self):
        Logger.log_info("Running scheduling")
        while not self.stopped:
            due_runnables = self.get_and_remove_due_runnables()

            while due_runnables:
                self.process_runnables(due_runnables)
                due_runnables = self.get_and_remove_due_runnables()

            with self.mutex:
                self.reschedule_required = False

            self.wait_until_runnable_is_due_or_reschedule_is_required()

        Logger.log_info("RunnableDispatcher shutdown.")

    def start(self):
        Logger.log_info("Python Starting RunnableDispatcher")
        with self.mutex:
            if not self.stopped:
                return False

            self.stopped = False
            Logger.log_info("Python Starting RunnableDispatcher")
            self.thread = threading.Thread(target=self.run_scheduling)
            self.thread.start()
            return True

    def stop(self):
        with self.mutex:
            if self.stopped:
                return False

            self.stopped = True

            self.condition_variable.notify()
        self.thread.join()
        return True

    def add_runnable(self, scheduled_runnable):
        Logger.log_info("Adding runnable 1")
        with self.mutex:
            Logger.log_info("Adding runnable 2")
            self.scheduled_runnables.append(scheduled_runnable)
            self.scheduled_runnables.sort(
                key=lambda x: x.schedule.get_execution_time()
            )
            Logger.log_info(
                f"Added runnable to RunnableDispatcher, total runnables now: {len(self.scheduled_runnables)}")

            self.reschedule_required = True
            self.condition_variable.notify_all()
            # This will lead to a wake up, so we can reschedule.
           

File: dispatch/java/runtime/adamma/c4dhi/claid/Module/Scheduling/RunnableDispatcher.java
/***************************************************************************
* Copyright (C) 2023 ETH Zurich
* CLAID: Closing the Loop on AI & Data Collection (https://claid.ethz.ch)
* Core AI & Digital Biomarker, Acoustic and Inflammatory Biomarkers (ADAMMA)
* Centre for Digital Health Interventions (c4dhi.org)
* 
* Authors: Patrick Langer
* 
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* 
*         http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
***************************************************************************/

package adamma.c4dhi.claid.Module.Scheduling;

import java.time.temporal.ChronoUnit;
import java.time.temporal.Temporal;
import java.time.temporal.TemporalUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import adamma.c4dhi.claid.Logger.Logger;
import adamma.c4dhi.claid_platform_impl.CLAID;
import adamma.c4dhi.claid.RemoteFunction.RemoteFunctionHandler;
import adamma.c4dhi.claid.RemoteFunction.RemoteFunction;
import adamma.c4dhi.claid.RuntimeType;
import adamma.c4dhi.claid.Runtime;

import java.util.concurrent.TimeUnit;
import java.util.Map;
import java.util.TreeMap;
import java.util.List;
import java.sql.Time;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;


import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.Duration;
import java.time.ZoneOffset;
import java.time.Instant;
public class RunnableDispatcher 
{
    private Map<LocalDateTime, ScheduledRunnable> scheduledRunnables = new TreeMap<>();
    private final Lock mutex = new ReentrantLock();
    private final Condition conditionVariable = mutex.newCondition();
    private boolean rescheduleRequired = false;
    private boolean running = false;

    private Thread scheduleThread;

    // Allows the RunnableDispatcher to make calls into the Middleware, e.g.,
    // to schedule device wake ups or request to keep the device awake.
    private RemoteFunctionHandler remoteFunctionHandler;

    private RemoteFunction<Void> middlewareFuncAcquireWakeLock;
    private RemoteFunction<Void> middlewareFuncReleaseLock;
    private RemoteFunction<Void> middlewareFuncScheduleDeviceWakeUpAt;

    private long localDateTimeToMilliseconds(LocalDateTime localDateTime)
    {
        long milliseconds = localDateTime.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli();
        return milliseconds;
    }

    private long getWaitDurationUntilNextRunnableIsDue() 
    {
        mutex.lock();
        try 
        {
            if (scheduledRunnables.isEmpty()) 
            {
                // Wait forever.
                return Long.MAX_VALUE;
            }

            LocalDateTime now = LocalDateTime.now();
            LocalDateTime nextExecutionTime = scheduledRunnables.keySet().iterator().next();

            long timeUntilNextRunnableIsDue = localDateTimeToMilliseconds(nextExecutionTime) - localDateTimeToMilliseconds(now);
            writeToLogFile("Calculating getWaitDurationUntilNextRunnableIsDue " + now.toString() + " " + nextExecutionTime.toString() + " " + timeUntilNextRunnableIsDue);

            // Make sure we do not return a negative duration.
            return timeUntilNextRunnableIsDue < 0 ? 0 : timeUntilNextRunnableIsDue;
        } 
        finally 
        {
            mutex.unlock();
        }
    }

    private void waitUntilRunnableIsDueOrRescheduleIsRequired() 
    {

        long waitTime = getWaitDurationUntilNextRunnableIsDue();



        mutex.lock();
        try {

             // Get the current LocalDateTime
            LocalDateTime now = LocalDateTime.now();

            // Define the offset in milliseconds
            long offsetMillis = waitTime; // Example: 1000 milliseconds (1 second)

            // Add the offset to the current time
            LocalDateTime newTime = now.plus(offsetMillis, ChronoUnit.MILLIS);
            // Define the formatter for the desired format
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy-HH:mm:ss");

            // Format the new time as a string
            String formattedTime = newTime.format(formatter);

            writeToLogFile("condition variable waiting for "  + waitTime + " milliseconds, waking up at " + formattedTime);



            middlewareScheduleDeviceWakeupAt(Long.valueOf(System.currentTimeMillis() + waitTime));

            conditionVariable.await(waitTime, TimeUnit.MILLISECONDS);

            writeToLogFile("Condition variable awoke");
        } catch (InterruptedException e) {
            writeToLogFile("Interrupted exception");
            Thread.currentThread().interrupt();

        } finally {
            writeToLogFile("Condition variable awoke unlocking");
            mutex.unlock();
        }
    }

    private void processRunnable(ScheduledRunnable scheduledRunnable) {
        if (!running) {
            return;
        }

        if (scheduledRunnable.isValid()) 
        {

            scheduledRunnable.run();
            


            /*if (scheduledRunnable.runnable.stopDispatcherAfterThisRunnable) {
                // No more runnables will be executed after this one!
                stop();
                System.out.println("STOPPED DISPATCHER!");
                return;
            }*/

            if (scheduledRunnable.schedule.doesRunnableHaveToBeRepeated())
             {
                writeToLogFile("Runnable needs to be repeated ");

                mutex.lock();
                try {

                    scheduledRunnable.schedule.updateExecutionTime();

                    LocalDateTime newExecutionTime = scheduledRunnable.schedule.getExecutionTime();
                    writeToLogFile("Runnable new execution time " + newExecutionTime.toString());


                    // Reinsert the runnable with the new scheduled execution time.
                    // Note that the runnable was removed from scheduledRunnables in the getAndRemoveDueRunnables() function.
                    scheduledRunnables.put(newExecutionTime, scheduledRunnable);
                } finally {
                    mutex.unlock();
                }
            }
        }
        else
        {
            Logger.logInfo("Runnable not valid.");
        }
    }

    private void processRunnables(List<ScheduledRunnable> runnables) {

        for (ScheduledRunnable runnable : runnables) {
            processRunnable(runnable);
        }
    }

    private List<ScheduledRunnable> getAndRemoveDueRunnables() {
        List<ScheduledRunnable> runnables = new ArrayList<>();
        long now = System.currentTimeMillis();

        mutex.lock();
        try {
            List<LocalDateTime> keysToRemove = new ArrayList<>();

            for (Map.Entry<LocalDateTime, ScheduledRunnable> entry : scheduledRunnables.entrySet()) {
                LocalDateTime dueTime = entry.getKey();

                long dueTimeMillis = localDateTimeToMilliseconds(dueTime);
                if (now >= dueTimeMillis) {
                    runnables.add(entry.getValue());
                    keysToRemove.add(dueTime);
                } else {
                    // Since the map is ordered in ascending order,
                    // all subsequent entries have a time
                    // which is greater than the time of the current entry.
                    // Hence, no need to keep iterating. We found the latest-most runnable
                    // which is due.
                    break;
                }
            }

            // If at least one runnable was added to runnables, that means we have to remove them from scheduledRunnables.
            for (LocalDateTime key : keysToRemove) {
                scheduledRunnables.remove(key);
            }
        } finally {
            mutex.unlock();
        }

        return runnables;
    }

    public void setRemoteFunctionHandler(RemoteFunctionHandler remoteFunctionHandler)
    {
        this.remoteFunctionHandler = remoteFunctionHandler;
        this.middlewareFuncAcquireWakeLock = 
            remoteFunctionHandler.mapRuntimeFunction(Runtime.MIDDLEWARE_CORE, "acquire_wakelock", Void.class, RuntimeType.class);

        this.middlewareFuncReleaseLock = 
            remoteFunctionHandler.mapRuntimeFunction(Runtime.MIDDLEWARE_CORE, "release_wakelock", Void.class, RuntimeType.class);

        this.middlewareFuncScheduleDeviceWakeUpAt = 
            remoteFunctionHandler.mapRuntimeFunction(Runtime.MIDDLEWARE_CORE, "schedule_device_wakeup_at", Void.class, RuntimeType.class, Long.class);
    }

    public boolean start() 
    {
        if(running)
        {
            return false;
        }
        running = true;
        scheduleThread = new Thread()
        {
            public void run()
            {
                runScheduling();
            }
        };
        scheduleThread.start();
        return true;
    }

    public boolean stop() {
        if(!running)
        {
            return false;
        }
        running = false;

        return true;
    }

    public void addRunnable(ScheduledRunnable runnable) {
        mutex.lock();
        try {
            LocalDateTime executionTime = runnable.schedule.getExecutionTime();
            
            // Ensure that we do not override an existing runnable.
            while(scheduledRunnables.containsKey(executionTime))
            {
                    Logger.logInfo("Runnables contains key");

                // Reschedule the runnable a bit later.
                // Plus 1000 ns = plus 1 ms
                executionTime = executionTime.plusNanos(1000);
            }

            scheduledRunnables.put(executionTime, runnable);
                Logger.logInfo("Runnables " + scheduledRunnables.size());
            // This will lead to a wake-up, so we can reschedule.
            rescheduleRequired = true;

            conditionVariable.signalAll();

        } finally {
            mutex.unlock();
        }

    }

    public void runScheduling() {
        List<ScheduledRunnable> dueRunnables = new ArrayList<>();
        while (running) {
            middlewareAcquireWakeLock();
            do {
                writeToLogFile("Scheduler woke up.");

                // While we process runnables, it is possible
                // that another runnable becomes due in the meantime.
                // Hence, we repeat this loop until there are no more
                // runnables that are due.
                dueRunnables = getAndRemoveDueRunnables();

                writeToLogFile("Scheduler due runnables: " + dueRunnables.size());

                processRunnables(dueRunnables);
            } while (!dueRunnables.isEmpty());
            rescheduleRequired = false;
            middlewareReleaseWakeLock();
            waitUntilRunnableIsDueOrRescheduleIsRequired();
        }
    }

    private FileOutputStream fos = null;
    void writeToLogFile(String data) 
    {
        Logger.logInfo(data);
        
        try {

            if(fos == null)
            {
                // Create a new File instance
                File file = new File(CLAID.getMediaDirPath(CLAID.getContext()) + "/java_scheduler_log.txt");

                // Use FileOutputStream to open the file in append mode
                fos = new FileOutputStream(file, true);
            }
            // Get the current date and time
            String dateTime = getCurrentDateTime();

            // Combine date, time, and data
            String entry = dateTime + " - " + data + "\n";

            // Write the data to the file
            fos.write(entry.getBytes());

            // Close the file output stream
        } catch (IOException e) {
            e.printStackTrace();
        } 
    }

    private static String getCurrentDateTime() {
        // Get the current date and time in the desired format
        SimpleDateFormat dateFormat = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss", Locale.getDefault());
        Date currentDate = new Date();
        return dateFormat.format(currentDate);
    }

    RuntimeType getRuntimeType()
    {
        RuntimeType.Builder type = RuntimeType.newBuilder();
        type.setRuntime(Runtime.RUNTIME_JAVA);
        return type.build();
    }

    private void middlewareAcquireWakeLock()
    {
        if(remoteFunctionHandler == null)
        {
            return;
        }
        middlewareFuncAcquireWakeLock.execute(getRuntimeType());
    }

    private void middlewareReleaseWakeLock()
    {
        if(remoteFunctionHandler == null)
        {
            return;
        }
        middlewareFuncReleaseLock.execute(getRuntimeType());
    }

    private void middlewareScheduleDeviceWakeupAt(Long milliseconds)
    {
        if(remoteFunctionHandler == null)
        {
            return;
        }
        middlewareFuncScheduleDeviceWakeUpAt.execute(getRuntimeType(), milliseconds);
    }
}
